sycl/plugins/esimd_cpu/esimd_libcm.hpp

```diff
@@ -0,0 +1,327 @@
+#include <thread>
+#include <vector>
+
+#define LAMBDA_WRAPPER_TMPL(ARGTYPE, TAG)                                      \
+  typedef std::function<void(const ARGTYPE &)> LambdaFunction_##TAG;           \
+                                                                               \
+  extern "C" struct LambdaWrapper_##TAG {                                      \
+    LambdaFunction_##TAG Func;                                                 \
+    std::vector<ARGTYPE> ArgVector;                                            \
+  };                                                                           \
+                                                                               \
+  template <typename LambdaTy>                                                 \
+  auto makeWrapper_##TAG(std::vector<ARGTYPE> &ArgVector, LambdaTy F) {        \
+    std::unique_ptr<LambdaWrapper_##TAG> Wrapper(new LambdaWrapper_##TAG);     \
+    Wrapper->Func = LambdaFunction_##TAG(F);                                   \
+    Wrapper->ArgVector = ArgVector;                                            \
+    return Wrapper;                                                            \
+  }                                                                            \
+                                                                               \
+  extern "C" inline void invokeLambda_##TAG(void *Wrapper) {                   \
+    auto *w = reinterpret_cast<LambdaWrapper_##TAG *>(Wrapper);                \
+    w->Func(w->ArgVector[cm_support::thread_local_idx()]);                     \
+  }
+
+#define _COMMA_ ,
+
+LAMBDA_WRAPPER_TMPL(sycl::id<3>, ID_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::id<2>, ID_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::id<1>, ID_1DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<3 _COMMA_ false>, ITEM_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<2 _COMMA_ false>, ITEM_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<1 _COMMA_ false>, ITEM_1DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<3 _COMMA_ true>, ITEM_OFFSET_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<2 _COMMA_ true>, ITEM_OFFSET_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::item<1 _COMMA_ true>, ITEM_OFFSET_1DIM)
+LAMBDA_WRAPPER_TMPL(sycl::nd_item<3>, NDITEM_3DIM)
+LAMBDA_WRAPPER_TMPL(sycl::nd_item<2>, NDITEM_2DIM)
+LAMBDA_WRAPPER_TMPL(sycl::nd_item<1>, NDITEM_1DIM)
+
+#undef LAMBDA_WRAPPER_TMPL
+#undef _COMMA_
+
+template <class KernelType, class KernelArgType, int __Dims__>
+class libCMBatch {
+private:
+  KernelType MKernel;
+  std::vector<KernelArgType> ArgVector;
+  std::vector<uint32_t> SpaceDim;
+  const uint32_t HWThreads = (uint32_t)std::thread::hardware_concurrency();
+
+  using IDBuilder = sycl::detail::Builder;
+
+public:
+  libCMBatch(KernelType Kernel) : MKernel(Kernel), SpaceDim{1, 1, 1} {}
+
+  // ID
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::id<__Dims__>>::value>::type
+  runIterationSpace(const sycl::range<__Dims__> &Range) {
+    sycl::detail::NDLoop<__Dims__>::iterate(
```

Comment:
This seems like a big performance issue. If the Range has e.g. 10^9 elements, ArgVector will take several gigs of memory. Please add FIXME and TODO tags. This needs to be addressed soon, I think.

Reply:
Will add comments. For your information, similar comments are already placed in source file to come in next stage.

Reply:
Right. I mentioned this before, just wanted to make sure it is not lost.

Reply:
In current CM_EMU's multi-threading implementation, all software threads are created/spawned with 'suspended' status at once, and each of them resumes whenever the number of active threads is below 'std::thread::hardware_concurrency()' value. This means CM creates and spawns 10^9 threads for the range with 10^9 elements at the beginning of kernel execution - even if 'ArgVector' is removed here. I'm not sure about Windows, but such massive thread spawning will trigger resource allocation error under Linux environment. Unless we are going to limit number of total threads, CM has to be modified to have 'batched' kernel launching mechanism.